# Single cell RNA-seq analysis with R
 - overview of preprocessing: from raw sequence reads to expression matrix
 - overview of popular tools and R packages for scRNAseq data analysis
 - scRNAseq data quality control 
 - cluster analysis
    - removal of undesired sources of variation
    - variable gene detection
    - dimensionality reduction
    - clustering
 - cell type identification
   - using known markers
   - using automatic classification algorithms
 - differential expression analysis
 - pseudotime analysis
 - if time permits: Integrating different datasets (CCA in Seurat)

## You will learn:
 - to assess the quality of scRNAseq data
 - to control batch effects and other unwanted variation
 - cell clustering and identification
 - gene expression analysis
 - choosing the tools for further analyses

## Prerequisites:
 - some experience in using R
 - understanding of the basic principles of single cell RNA-seq experiments